glib-macros/src/closure_old.rs

```diff
@@ -0,0 +1,300 @@
+            _ => Default::default(),
+        }
+    }
+
+    fn inner_before_tokens(&self, crate_ident: &TokenStream) -> TokenStream {
+        let alias = self.alias();
+        match self.kind {
+            CaptureKind::Watch => {
+                quote! {
+                    let #alias = unsafe { #alias.borrow() };
+                    let #alias = ::core::convert::AsRef::as_ref(&#alias);
+                }
+            }
+            CaptureKind::WeakAllowNone => quote! {
+                let #alias = #crate_ident::clone::Upgrade::upgrade(&#alias);
+            },
+            _ => Default::default(),
+        }
+    }
+}
+
+impl syn::parse::Parse for CaptureKind {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        input.parse::<Token![@]>()?;
+        let mut idents = TokenStream::new();
+        idents.append(input.call(syn::Ident::parse_any)?);
+        while input.peek(Token![-]) {
+            input.parse::<Token![-]>()?;
+            idents.append(input.call(syn::Ident::parse_any)?);
+        }
+        let keyword = idents
+            .clone()
+            .into_iter()
+            .map(|i| i.to_string())
+            .collect::<Vec<_>>()
+            .join("-");
+        match keyword.as_str() {
+            "strong" => Ok(CaptureKind::Strong),
+            "watch" => Ok(CaptureKind::Watch),
+            "weak-allow-none" => Ok(CaptureKind::WeakAllowNone),
+            "to-owned" => Ok(CaptureKind::ToOwned),
+            k => Err(syn::Error::new(
+                idents.span(),
+                format!("Unknown keyword `{}`, only `watch`, `weak-allow-none`, `to-owned` and `strong` are allowed",
+                k),
+            )),
+        }
+    }
+}
+
+impl syn::parse::Parse for Capture {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let start = input.span();
+        let kind = input.parse()?;
+        let mut name = TokenStream::new();
+        name.append(input.call(syn::Ident::parse_any)?);
+        while input.peek(Token![.]) {
+            input.parse::<Token![.]>()?;
+            name.append(proc_macro2::Punct::new('.', proc_macro2::Spacing::Alone));
+            name.append(input.call(syn::Ident::parse_any)?);
+        }
+        let alias = if input.peek(Token![as]) {
+            input.parse::<Token![as]>()?;
+            input.parse()?
+        } else {
+            None
+        };
+        if alias.is_none() {
+            if name.to_string() == "self" {
+                return Err(syn::Error::new_spanned(
+                    name,
+                    "Can't use `self` as variable name. Try storing it in a temporary variable or \
+                    rename it using `as`.",
+                ));
+            }
+            if name.to_string().contains('.') {
+                return Err(syn::Error::new(
+                    name.span(),
+                    format!(
+                        "`{}`: Field accesses are not allowed as is, you must rename it!",
```

Comment:
You could use (not a blocker, just a style thing, I personally have no preference, just wanted to mention it):

```suggestion
                        "`{name}`: Field accesses are not allowed as is, you must rename it!",
```

Reply:
Please ignore, old file.